File: implementations/typescript/tests/tools/azure-firewall.test.ts
import { describe, expect, test } from 'bun:test';
import { 
  FirewallRuleSchema, 
  createAzureFirewallHandler,
  getAzureFirewallSchema 
} from '../../tools/azure-firewall';

describe('FirewallRuleSchema', () => {
  test('accepts valid input with all fields', () => {
    const result = FirewallRuleSchema.safeParse({
      team: 'platform',
      source: '10.0.0.0/8',
      destination: '192.168.1.100',
      port: 443,
      existing_yaml: '',
    });
    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.team).toBe('platform');
      expect(result.data.port).toBe(443);
    }
  });

  test('accepts input without port (uses default)', () => {
    const result = FirewallRuleSchema.safeParse({
      team: 'platform',
      source: '10.0.0.1',
      destination: '192.168.1.100',
      existing_yaml: '',
    });
    expect(result.success).toBe(true);
    if (result.success) {
      expect(result.data.port).toBe(443);
    }
  });

  test('rejects invalid port', () => {
    const result = FirewallRuleSchema.safeParse({
      team: 'platform',
      source: '10.0.0.1',
      destination: '192.168.1.100',
      port: 70000,
      existing_yaml: '',
    });
    expect(result.success).toBe(false);
  });

  test('rejects missing required fields', () => {
    const result = FirewallRuleSchema.safeParse({
      team: 'platform',
    });
    expect(result.success).toBe(false);
  });

  test('rejects invalid IP format via regex pattern', () => {
    const result = FirewallRuleSchema.safeParse({
      team: 'platform',
      source: 'not-an-ip',
      destination: '192.168.1.100',
      existing_yaml: '',
    });
    expect(result.success).toBe(false);
  });

  test('accepts CIDR notation', () => {
    const result = FirewallRuleSchema.safeParse({
      team: 'platform',
      source: '10.0.0.0/24',
      destination: '192.168.0.0/16',
      existing_yaml: '',
    });
    expect(result.success).toBe(true);
  });
});

describe('createAzureFirewallHandler - Content Return Pattern', () => {
  test('creates new YAML content when existing_yaml is empty', async () => {
    const handler = createAzureFirewallHandler();
    const result = await handler(
      {
        team: 'platform',
        source: '10.0.0.1',
        destination: '192.168.1.1',
        port: 443,
        existing_yaml: '',
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBeUndefined();
    
    const response = JSON.parse(result.content[0]?.text || '{}');
    expect(response.action).toBe('created');
    expect(response.filename).toBe('azure-firewall-rules.yaml');
    expect(response.yaml_content).toContain('team: platform');
    expect(response.yaml_content).toContain('source: 10.0.0.1');
    expect(response.message).toContain('created');
  });

  test('updates existing YAML content', async () => {
    const handler = createAzureFirewallHandler();
    
    const existingYaml = `rules:
- team: existing-team
  source: 1.1.1.1
  destination: 2.2.2.2
  port: 80
`;
    
    const result = await handler(
      {
        team: 'new-team',
        source: '10.0.0.1',
        destination: '192.168.1.1',
        port: 443,
        existing_yaml: existingYaml,
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBeUndefined();
    
    const response = JSON.parse(result.content[0]?.text || '{}');
    expect(response.action).toBe('updated');
    expect(response.yaml_content).toContain('existing-team');
    expect(response.yaml_content).toContain('new-team');
  });

  test('detects duplicate rules and returns duplicate_detected action', async () => {
    const handler = createAzureFirewallHandler();
    
    const existingYaml = `rules:
- team: platform
  source: 10.0.0.1
  destination: 192.168.1.1
  port: 443
`;
    
    const result = await handler(
      {
        team: 'platform',
        source: '10.0.0.1',
        destination: '192.168.1.1',
        port: 443,
        existing_yaml: existingYaml,
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBeUndefined();
    
    const response = JSON.parse(result.content[0]?.text || '{}');
    expect(response.action).toBe('duplicate_detected');
    expect(response.message).toContain('already exists');
  });

  test('rejects invalid source IP', async () => {
    const handler = createAzureFirewallHandler();
    const result = await handler(
      {
        team: 'platform',
        source: 'invalid-ip',
        destination: '192.168.1.1',
        port: 443,
        existing_yaml: '',
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBe(true);
    expect(result.content[0]?.text).toContain('Invalid source IP or CIDR');
  });

  test('rejects invalid destination IP', async () => {
    const handler = createAzureFirewallHandler();
    const result = await handler(
      {
        team: 'platform',
        source: '10.0.0.1',
        destination: 'not-an-ip',
        port: 443,
        existing_yaml: '',
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBe(true);
    expect(result.content[0]?.text).toContain('Invalid destination IP or CIDR');
  });

  test('returns parse error for invalid existing_yaml', async () => {
    const handler = createAzureFirewallHandler();
    const result = await handler(
      {
        team: 'platform',
        source: '10.0.0.1',
        destination: '192.168.1.1',
        port: 443,
        existing_yaml: 'this: is: not: valid: yaml: [[[',
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBe(true);
    expect(result.content[0]?.text).toContain('Failed to parse existing_yaml');
  });

  test('handles existing YAML without rules key', async () => {
    const handler = createAzureFirewallHandler();
    const result = await handler(
      {
        team: 'platform',
        source: '10.0.0.1',
        destination: '192.168.1.1',
        port: 443,
        existing_yaml: 'some_other_key: value',
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBeUndefined();
    
    const response = JSON.parse(result.content[0]?.text || '{}');
    expect(response.action).toBe('created');
    expect(response.yaml_content).toContain('rules:');
  });

  test('uses default port 443 when not specified', async () => {
    const handler = createAzureFirewallHandler();
    // Port defaults via schema, but handler receives the defaulted value
    const result = await handler(
      {
        team: 'platform',
        source: '10.0.0.1',
        destination: '192.168.1.1',
        port: 443, // Schema default applied before handler
        existing_yaml: '',
      },
      { executionId: 'test-123' }
    );

    expect(result.isError).toBeUndefined();
    
    const response = JSON.parse(result.content[0]?.text || '{}');
    expect(response.yaml_content).toContain('port: 443');
  });
});

describe('getAzureFirewallSchema', () => {
  test('returns valid JSON schema', () => {
    const schema = getAzureFirewallSchema();
    const parsed = JSON.parse(schema);
    
    expect(parsed.$schema).toBe('https://json-schema.org/draft/2020-12/schema');
    expect(parsed.type).toBe('object');
    expect(parsed.properties.rules).toBeDefined();
    expect(parsed.required).toContain('rules');
  });

  test('includes pattern validation for IP addresses', () => {
    const schema = getAzureFirewallSchema();
    const parsed = JSON.parse(schema);
    
    expect(parsed.properties.rules.items.properties.source.pattern).toBeDefined();
    expect(parsed.properties.rules.items.properties.destination.pattern).toBeDefined();
  });
});

File: implementations/go/internal/tools/azure_firewall_test.go
package tools

import (
	"context"
	"encoding/json"
	"strings"
	"testing"

	"github.com/modelcontextprotocol/go-sdk/mcp"
)

func TestNewAzureFirewallRules_ContentReturnPattern_CreateNew(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	input := FirewallRuleInput{
		Team:         "platform",
		Source:       "10.0.0.1",
		Destination:  "192.168.1.1",
		Port:         443,
		ExistingYAML: "",
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.IsError {
		t.Fatalf("Expected no result error, got true")
	}

	// Parse the response JSON
	var response FirewallToolResponse
	textContent := result.Content[0].(*mcp.TextContent)
	if err := json.Unmarshal([]byte(textContent.Text), &response); err != nil {
		t.Fatalf("Failed to parse response JSON: %v", err)
	}

	if response.Action != ActionCreated {
		t.Errorf("Expected action 'created', got %s", response.Action)
	}

	if response.Filename != "azure-firewall-rules.yaml" {
		t.Errorf("Expected filename 'azure-firewall-rules.yaml', got %s", response.Filename)
	}

	if !strings.Contains(response.YAMLContent, "team: platform") {
		t.Errorf("Expected yaml_content to contain 'team: platform', got %s", response.YAMLContent)
	}

	if !strings.Contains(response.YAMLContent, "source: 10.0.0.1") {
		t.Errorf("Expected yaml_content to contain 'source: 10.0.0.1', got %s", response.YAMLContent)
	}

	if !strings.Contains(response.Message, "created") {
		t.Errorf("Expected message to contain 'created', got %s", response.Message)
	}
}

func TestNewAzureFirewallRules_ContentReturnPattern_UpdateExisting(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	existingYAML := `rules:
- team: existing-team
  source: 1.1.1.1
  destination: 2.2.2.2
  port: 80
`

	input := FirewallRuleInput{
		Team:         "new-team",
		Source:       "10.0.0.1",
		Destination:  "192.168.1.1",
		Port:         443,
		ExistingYAML: existingYAML,
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.IsError {
		t.Fatalf("Expected no result error, got true")
	}

	var response FirewallToolResponse
	textContent := result.Content[0].(*mcp.TextContent)
	if err := json.Unmarshal([]byte(textContent.Text), &response); err != nil {
		t.Fatalf("Failed to parse response JSON: %v", err)
	}

	if response.Action != ActionUpdated {
		t.Errorf("Expected action 'updated', got %s", response.Action)
	}

	if !strings.Contains(response.YAMLContent, "existing-team") {
		t.Errorf("Expected yaml_content to contain 'existing-team', got %s", response.YAMLContent)
	}

	if !strings.Contains(response.YAMLContent, "new-team") {
		t.Errorf("Expected yaml_content to contain 'new-team', got %s", response.YAMLContent)
	}
}

func TestNewAzureFirewallRules_ContentReturnPattern_DuplicateDetected(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	existingYAML := `rules:
- team: platform
  source: 10.0.0.1
  destination: 192.168.1.1
  port: 443
`

	input := FirewallRuleInput{
		Team:         "platform",
		Source:       "10.0.0.1",
		Destination:  "192.168.1.1",
		Port:         443,
		ExistingYAML: existingYAML,
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.IsError {
		t.Fatalf("Expected no result error for duplicate detection, got true")
	}

	var response FirewallToolResponse
	textContent := result.Content[0].(*mcp.TextContent)
	if err := json.Unmarshal([]byte(textContent.Text), &response); err != nil {
		t.Fatalf("Failed to parse response JSON: %v", err)
	}

	if response.Action != ActionDuplicateDetected {
		t.Errorf("Expected action 'duplicate_detected', got %s", response.Action)
	}

	if !strings.Contains(response.Message, "already exists") {
		t.Errorf("Expected message to contain 'already exists', got %s", response.Message)
	}
}

func TestNewAzureFirewallRules_ContentReturnPattern_InvalidSourceIP(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	input := FirewallRuleInput{
		Team:         "platform",
		Source:       "invalid-ip",
		Destination:  "192.168.1.1",
		Port:         443,
		ExistingYAML: "",
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if !result.IsError {
		t.Fatalf("Expected result.IsError to be true for invalid source IP")
	}

	textContent := result.Content[0].(*mcp.TextContent)
	if !strings.Contains(textContent.Text, "Invalid source IP or CIDR") {
		t.Errorf("Expected error message about invalid source IP, got %s", textContent.Text)
	}
}

func TestNewAzureFirewallRules_ContentReturnPattern_InvalidDestinationIP(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	input := FirewallRuleInput{
		Team:         "platform",
		Source:       "10.0.0.1",
		Destination:  "not-an-ip",
		Port:         443,
		ExistingYAML: "",
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if !result.IsError {
		t.Fatalf("Expected result.IsError to be true for invalid destination IP")
	}

	textContent := result.Content[0].(*mcp.TextContent)
	if !strings.Contains(textContent.Text, "Invalid destination IP or CIDR") {
		t.Errorf("Expected error message about invalid destination IP, got %s", textContent.Text)
	}
}

func TestNewAzureFirewallRules_ContentReturnPattern_InvalidYAML(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	input := FirewallRuleInput{
		Team:         "platform",
		Source:       "10.0.0.1",
		Destination:  "192.168.1.1",
		Port:         443,
		ExistingYAML: "this: is: not: valid: yaml: [[[",
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if !result.IsError {
		t.Fatalf("Expected result.IsError to be true for invalid YAML")
	}

	textContent := result.Content[0].(*mcp.TextContent)
	if !strings.Contains(textContent.Text, "Failed to parse existing_yaml") {
		t.Errorf("Expected error message about parsing YAML, got %s", textContent.Text)
	}
}

func TestNewAzureFirewallRules_ContentReturnPattern_ExistingYAMLWithoutRulesKey(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	input := FirewallRuleInput{
		Team:         "platform",
		Source:       "10.0.0.1",
		Destination:  "192.168.1.1",
		Port:         443,
		ExistingYAML: "some_other_key: value",
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.IsError {
		t.Fatalf("Expected no result error, got true")
	}

	var response FirewallToolResponse
	textContent := result.Content[0].(*mcp.TextContent)
	if err := json.Unmarshal([]byte(textContent.Text), &response); err != nil {
		t.Fatalf("Failed to parse response JSON: %v", err)
	}

	if response.Action != ActionCreated {
		t.Errorf("Expected action 'created' for YAML without rules key, got %s", response.Action)
	}

	if !strings.Contains(response.YAMLContent, "rules:") {
		t.Errorf("Expected yaml_content to contain 'rules:', got %s", response.YAMLContent)
	}
}

func TestNewAzureFirewallRules_ContentReturnPattern_DefaultPort(t *testing.T) {
	ctx := context.Background()
	req := &mcp.CallToolRequest{}

	input := FirewallRuleInput{
		Team:         "platform",
		Source:       "10.0.0.1",
		Destination:  "192.168.1.1",
		Port:         0, // 0 means not provided, should default to 443
		ExistingYAML: "",
	}

	handler := NewAzureFirewallRulesHandler()
	result, _, err := handler(ctx, req, input)
	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.IsError {
		t.Fatalf("Expected no result error, got true")
	}

	var response FirewallToolResponse
	textContent := result.Content[0].(*mcp.TextContent)
	if err := json.Unmarshal([]byte(textContent.Text), &response); err != nil {
		t.Fatalf("Failed to parse response JSON: %v", err)
	}

	if !strings.Contains(response.YAMLContent, "port: 443") {
		t.Errorf("Expected yaml_content to contain 'port: 443', got %s", response.YAMLContent)
	}
}

func TestGetAzureFirewallSchema_ValidJSONSchema(t *testing.T) {
	schemaStr := GetAzureFirewallSchema()

	var schema map[string]any
	if err := json.Unmarshal([]byte(schemaStr), &schema); err != nil {
		t.Fatalf("Failed to parse schema JSON: %v", err)
	}

	if schema["$schema"] != "https://json-schema.org/draft/2020-12/schema" {
		t.Errorf("Expected $schema to be draft/2020-12, got %v", schema["$schema"])
	}

	if schema["type"] != "object" {
		t.Errorf("Expected type to be 'object', got %v", schema["type"])
	}

	properties, ok := schema["properties"].(map[string]any)
	if !ok {
		t.Fatalf("Expected properties to be a map")
	}

	if _, exists := properties["rules"]; !exists {
		t.Errorf("Expected properties to contain 'rules'")
	}

	required, ok := schema["required"].([]any)
	if !ok {
		t.Fatalf("Expected required to be an array")
	}

	hasRules := false
	for _, r := range required {
		if r == "rules" {
			hasRules = true
			break
		}
	}
	if !hasRules {
		t.Errorf("Expected required to contain 'rules'")
	}
}

func TestGetAzureFirewallSchema_IncludesPatternValidation(t *testing.T) {
	schemaStr := GetAzureFirewallSchema()

	var schema map[string]any
	if err := json.Unmarshal([]byte(schemaStr), &schema); err != nil {
		t.Fatalf("Failed to parse schema JSON: %v", err)
	}

	properties := schema["properties"].(map[string]any)
	rules := properties["rules"].(map[string]any)
	items := rules["items"].(map[string]any)
	itemProps := items["properties"].(map[string]any)

	source := itemProps["source"].(map[string]any)
	if _, exists := source["pattern"]; !exists {
		t.Errorf("Expected source to have pattern validation")
	}

	destination := itemProps["destination"].(map[string]any)
	if _, exists := destination["pattern"]; !exists {
		t.Errorf("Expected destination to have pattern validation")
	}
}
